File: analytics/src/main/java/io/appmetrica/analytics/impl/utils/UnlockedUserStateProvider.java
package io.appmetrica.analytics.impl.utils;

import android.content.Context;
import android.os.Build;
import android.os.UserManager;
import androidx.annotation.NonNull;
import io.appmetrica.analytics.coreapi.internal.backport.FunctionWithThrowable;
import io.appmetrica.analytics.coreutils.internal.system.SystemServiceUtils;

public class UnlockedUserStateProvider {

    public boolean isUserUnlocked(@NonNull Context context) {
        boolean unlocked = true;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            UserManager userManager = context.getSystemService(UserManager.class);
            unlocked = SystemServiceUtils.accessSystemServiceSafelyOrDefault(
                userManager,
                "detect unlocked user state",
                "User manager",
                true,
                new FunctionWithThrowable<UserManager, Boolean>() {
                    @Override
                    public Boolean apply(@NonNull UserManager input) throws Throwable {
                        return input.isUserUnlocked();
                    }
                }
            );
        }
        return unlocked;
    }

}

File: analytics/src/main/java/io/appmetrica/analytics/impl/utils/UnlockedUserStateProvider.kt
package io.appmetrica.analytics.impl.utils

import android.content.Context
import android.os.Build
import android.os.UserManager
import androidx.annotation.RequiresApi
import io.appmetrica.analytics.coreutils.internal.AndroidUtils.isApiAchieved
import io.appmetrica.analytics.coreutils.internal.system.SystemServiceUtils

internal class UnlockedUserStateProvider {
    @SuppressWarnings("NewApi")
    fun isUserUnlocked(context: Context): Boolean = if (isApiAchieved(Build.VERSION_CODES.N)) {
        isUserUnlockedStateSinceN(context)
    } else true

    @RequiresApi(Build.VERSION_CODES.N)
    fun isUserUnlockedStateSinceN(context: Context): Boolean {
        val userManager = context.getSystemService(UserManager::class.java)
        return SystemServiceUtils.accessSystemServiceSafelyOrDefault<UserManager?, Boolean?>(
            userManager,
            "detect unlocked user state",
            "User manager",
            true
        ) { input -> input.isUserUnlocked } ?: true
    }
}
